File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int toSolve(int n, int[] temp){
        int prev2 = 0, prev = temp[0];

        for(int i=1;i<n;i++){

            int take = temp[i];
            if(i - 2 >= 0) take += prev2;
            int nonTake = 0 + prev;

            int cur = Math.max(take, nonTake);
            prev2 = prev;
            prev = cur;
        }

        return prev;
    }

    public int rob(int[] nums) {
        int n = nums.length;
        if(n == 1) return nums[0];

        int[] temp = new int[n-1];
        for(int i=1;i<n;i++){
            temp[i-1] = nums[i];
        }

        int withoutFirst = toSolve(n-1, temp);

        for(int i=0;i<n-1;i++){
            temp[i] = nums[i];
        }

        int withoutLast= toSolve(n-1, temp);

        return Math.max(withoutFirst, withoutLast);


    }
}

File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    
    int solve(int ind, vector<int> &nums, vector<int> &dp){
        if(ind == 0) return nums[ind];
        if(ind < 0) return 0;
        if(dp[ind] != -1) return dp[ind];


        int take = nums[ind] + solve(ind-2, nums, dp);

        int notTake = 0 + solve(ind-1, nums, dp);

        return dp[ind] = max(take, notTake);
    }

    int rob(vector<int>& nums) {
        if(nums.size() == 1) return nums[0];
        else if(nums.size() == 2) return max(nums[0], nums[1]);

        int last_element = nums[nums.size() - 1];

        nums.assign(nums.begin() , nums.begin() + nums.size() -1);

        // nums[0...n-2]

        vector<int> dp(nums.size(), -1);
        int firstAns = solve(nums.size()-1, nums, dp);

        // nums[1...n-1]
        nums.erase(nums.begin());
        nums.push_back(last_element);

        fill(dp.begin(), dp.end(), -1);


        int secondAns = solve(nums.size()-1, nums, dp);

        return max(firstAns, secondAns);

        return 0;
    }
};